planner/core/logical_plans.go

```diff
@@ -902,6 +909,29 @@ type FrameBound struct {
 	CmpFuncs []expression.CompareFunc
 }
 
+// Equal checks whether two FrameBounds are equal.
+func (bound *FrameBound) Equal(ctx sessionctx.Context, other *FrameBound) bool {
+	if bound.Type != other.Type || bound.UnBounded != other.UnBounded || bound.Num != other.Num {
+		return false
+	}
+	if len(bound.CalcFuncs) != len(other.CalcFuncs) || len(bound.CmpFuncs) != len(other.CmpFuncs) {
+		return false
+	}
+	for i := range bound.CalcFuncs {
+		if !bound.CalcFuncs[i].Equal(ctx, other.CalcFuncs[i]) {
+			return false
+		}
+	}
+	for i := range bound.CmpFuncs {
+		v1 := reflect.ValueOf(&bound.CmpFuncs[i]).Elem()
+		v2 := reflect.ValueOf(&other.CmpFuncs[i]).Elem()
+		if v1.Pointer() != v2.Pointer() {
+			return false
+		}
+	}
```

Comment:
I'm not sure about this method for the comparison of `CmpFuncs`. Any ideas? cc @winoros

Reply:
It's ok. CmpFuncs is generated via `GetCmpFunction`. We can use function pointer to check whether it's the same one.